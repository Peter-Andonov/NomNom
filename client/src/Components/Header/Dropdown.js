import React from 'react';
import styled from 'styled-components';
import StyledLink from './Link';

export default function DropdownMenu() {

    const Container = styled.div`
    position: absolute;
    top: 100px;
    width: auto;
    min-width: 175px;
    display: flex;
    flex-direction: column;
    align-items: center;
    background: #333333;
    border: 1px solid #474a4d;
    border-radius: 5px;
    `

    return (
        <Container>
            <StyledLink to={'/admin'} text={'Admin'} />
            <StyledLink to={'/profile'} text={'Profile'} />
            <StyledLink to={'/logout'} text={'Logout'} />
        </Container>
    );
}