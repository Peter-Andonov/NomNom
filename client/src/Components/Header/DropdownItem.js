import React from 'react';
import { Link } from 'react-router-dom';
import styled from 'styled-components';

export default function DropdownItem(props) {

    const StyledLink = styled(Link)`
    margin: 0.5rem;
    color: white;
    text-decoration: none;
    size: 0.8rem;
    &:hover {
        color: rgba(237, 71, 59);
        cursor: pointer;
    }
    `

    return (
        <StyledLink to={props.url}>{props.text}</StyledLink>
    );
}