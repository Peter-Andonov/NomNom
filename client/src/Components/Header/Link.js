import React from 'react';
import { Link } from 'react-router-dom';
import styled from 'styled-components';


const StyledLink = styled(Link)`
    margin-left: 1rem;
    margin-right: 1rem;
    color: white;
    text-decoration: none;
    size: 0.8rem;
    &:hover {
        color: rgba(237, 71, 59);
        cursor: pointer;
    }
`;


const NavLink = (props) => {

    return (
        <StyledLink to={props.to}>{props.text}</StyledLink>
    );
};


export default NavLink;