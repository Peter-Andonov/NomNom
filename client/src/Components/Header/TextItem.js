import React from 'react';
import styled from 'styled-components';


const Div = styled.div`
    margin-left: 1rem;
    margin-right: 1rem;
    color: white;
    size: 0.8rem;
    &:hover {
        color: rgba(237, 71, 59);
        cursor: pointer;
    }
`;


const TextItem = (props) => {
    return(
        <Div onClick={props.action}>{props.label}</Div>
    );
};


export default TextItem;