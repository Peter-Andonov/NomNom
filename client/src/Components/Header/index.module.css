.header {
    top: 0;
    width: 100%;
    height: 8vh;
    min-height: 4rem;
    position: fixed;
    background-color: var(--bg-primary);
}

.header-ul {
    list-style: none;
    padding: 0;
    margin: 0;
    display: flex;
    align-items: center;
    height: 100%;
}

.spacer {
    flex: 1
}

.header-item {
    height: 100%;
}

.header-link {
    display: flex;
    align-items: center;
    color: var(--text-primary);
    height: 100%;
    text-decoration: none;
    filter: grayscale(100%) opacity(0.7);
    transition: var(--transition-speed);
}

.header-link:hover {
    filter: grayscale(0%) opacity(1);
    background: var(--bg-secondary);
    color: var(--text-secondary);
 }

.link-text {
    margin: 1rem;
    color: var(--text-primary);
}

.profile-pic {
    margin-left: 1rem;
    height: 80%;
    max-height: 3rem;
    border-radius: 50%;
}

.logo {
    font-weight: bold;
    text-align: center;
    font-size: 2rem;
    letter-spacing: 0.3ch;
    height: 100%;
    padding: 1rem;
}

.logo-link {
    display: flex;
    align-items: center;
    height: 100%;
    text-decoration: none;
}

.logo-text {
    margin: 1rem;
    color: #ff8533;
}

.logo-svg {
    color: #993300;
}