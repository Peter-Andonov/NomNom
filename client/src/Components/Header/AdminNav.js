import React, { useState } from 'react';
import styled from 'styled-components';
import TextItem from './TextItem';


const Container = styled.div`
    color: white;
    text-decoration: none;
    display: flex;
    flex-direction: row;
    align-items: center;
    &:hover {
        cursor: pointer;
    }
`;

const AdminNav = (props) => {

    const [open, setOpen] = useState(false);

    const toggleOpen = () => {
        setOpen(!open)
    };

    return (
        <Container onClick={toggleOpen}>
            <TextItem label={"Admin"} />
            {open && props.children}
        </Container>
    );
};


export default AdminNav;