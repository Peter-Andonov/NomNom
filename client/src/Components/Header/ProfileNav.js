import React, { useState, useContext } from 'react';
import styled from 'styled-components';
import UserContext from '../../Context';
import TextItem from './TextItem';


const Container = styled.div`
    margin: 1rem;
    color: white;
    text-decoration: none;
    display: flex;
    flex-direction: row;
    align-items: center;
    &:hover {
        cursor: pointer;
    }
`;

const Avatar = styled.img`
    margin-right: 1rem;
    width: 4rem;
    height: 4rem;
    border-radius: 50%;
    object-fit: cover;
`;


const ProfileNav = (props) => {

    const [open, setOpen] = useState(false);

    const toggleOpen = () => {
        setOpen(!open)
    };

    const userContext = useContext(UserContext);
    const displayName = userContext.user.firstName ? `${userContext.user.firstName} ${userContext.user.lastName}` : userContext.user.email;
    const displayPicture = userContext.user.profilePicUrl || 'https://res.cloudinary.com/nomnomapp/image/upload/v1596641498/Images/x5svb4l6kwcyiinazfqt.png';
    
    return (
        <Container onClick={toggleOpen}>
            <Avatar src={displayPicture} alt='Avatar' />
            <TextItem label={displayName} />
            {open && props.children}
        </Container>
    );

};


export default ProfileNav;