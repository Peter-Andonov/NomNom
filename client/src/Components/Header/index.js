import React, { useContext } from 'react';
import { useHistory } from "react-router-dom";
import UserContext from '../../Context';
import styled from 'styled-components';
import ProfileNav from './ProfileNav';
import AdminNav from './AdminNav';
import Dropdown from './Dropdown';
import StyledLink from './Link';
import TextItem from './TextItem';


const Head = styled.header`
    top: 0;
    width: 100%;
    height: 100px;
    position: fixed;
    background-image: linear-gradient(rgba(0,0,0, .9),rgba(245,245,220, 0));
`;

const Navigation = styled.nav`
    height: 100%;
    display: flex;
    flex-direction: row;
    align-items: center;
`;

const Spacer = styled.div`
    flex: 1;
`;

const Header = () => {

    const userContext = useContext(UserContext);
    const history = useHistory();
    const isAdmin = userContext.user ? userContext.user.role === 'admin' : false;

    const logOut = (e) => {
        userContext.logOut();
        history.push('/');
    };


    return (
        <Head>
            <Navigation>
                <StyledLink to={'/'} text={'Home'} />
                <StyledLink to={'/recipes'} text={'Recipes'} />
                <StyledLink to={'/ingredients'} text={'Ingredients'} />
                <StyledLink to={'/articles'} text={'Articles'} />
                {isAdmin &&
                    <AdminNav >
                        <Dropdown>
                            <StyledLink to={'/admin'} text={'Dashboard'} />
                            <StyledLink to={'/create/recipe'} text={'Create Recipe'} />
                            <StyledLink to={'/create/ingredient'} text={'Create Ingredient'} />
                            <StyledLink to={'/create/article'} text={'Create Article'} />
                            <StyledLink to={'/create/unit'} text={'Create Unit'} />
                        </Dropdown>
                    </AdminNav>}
                <Spacer />
                {!userContext.loggedIn &&
                    <div>
                        <StyledLink to={'/login'} text={'Login'} />
                        <StyledLink to={'/register'} text={'Register'} />
                    </div>}
                {userContext.loggedIn &&
                    <ProfileNav>
                        <Dropdown>
                            {!isAdmin && <StyledLink to={'/favourites'} text={'Favourites'} />}
                            <StyledLink to={`/profile`} text={'Profile'} />
                            <TextItem action={logOut} label={'Logout'} />
                        </Dropdown>
                    </ProfileNav>}
            </Navigation>
        </Head>
    );
};


export default Header;