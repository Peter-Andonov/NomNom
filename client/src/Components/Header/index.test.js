import React from 'react'
import { render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import TestSetup from '../../Utils/TestUtils/test-utils';


import Header from './index';

jest.mock('./ProfileNav.js', () => 'ProfileNav');

describe('Header component', () => {
    it("renders guest navigation", () => {
        const { asFragment } = render(
            <TestSetup value={{
                loggedIn: false,
                user: null
            }}>
                <Header />
            </TestSetup>);
        expect(asFragment()).toMatchSnapshot();
    });

    it("renders user navigation", () => {
        const { asFragment } = render(
            <TestSetup value={{
                loggedIn: true,
                user: {
                    _id: '1234',
                    email: "john.doe@gmail.com",
                    role: 'user',
                    firstName: undefined,
                    lastName: undefined,
                    profilePicUrl: '',
                    favouriteRecipes: []
                }}}>
                <Header />
            </TestSetup>);
        expect(asFragment()).toMatchSnapshot();
    });

    it("renders admin navigation", () => {
        const { asFragment } = render(
            <TestSetup value={{
                loggedIn: true,
                user: {
                    _id: '1234',
                    email: "john.doe@gmail.com",
                    role: 'admin',
                    firstName: undefined,
                    lastName: undefined,
                    profilePicUrl: '',
                    favouriteRecipes: []
                }}}>
                <Header />
            </TestSetup>);
        expect(asFragment()).toMatchSnapshot();
    });
});