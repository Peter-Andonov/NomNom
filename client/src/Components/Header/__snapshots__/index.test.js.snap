// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component renders admin navigation 1`] = `
<DocumentFragment>
  <header
    class="sc-AxhUy koLwTO"
  >
    <nav
      class="sc-AxgMl fOKyzp"
    >
      <a
        class="sc-AxhCb eZTMiI"
        href="/"
      >
        Home
      </a>
      <a
        class="sc-AxhCb eZTMiI"
        href="/recipes"
      >
        Recipes
      </a>
      <a
        class="sc-AxhCb eZTMiI"
        href="/ingredients"
      >
        Ingredients
      </a>
      <a
        class="sc-AxhCb eZTMiI"
        href="/articles"
      >
        Articles
      </a>
      <div
        class="sc-AxirZ GpSyG"
      >
        <div
          class="sc-AxjAm cxNAeo"
        >
          Admin
        </div>
      </div>
      <div
        class="sc-AxheI jYYRbo"
      />
      <div />
    </nav>
  </header>
</DocumentFragment>
`;

exports[`Header component renders guest navigation 1`] = `
<DocumentFragment>
  <header
    class="sc-AxhUy koLwTO"
  >
    <nav
      class="sc-AxgMl fOKyzp"
    >
      <a
        class="sc-AxhCb eZTMiI"
        href="/"
      >
        Home
      </a>
      <a
        class="sc-AxhCb eZTMiI"
        href="/recipes"
      >
        Recipes
      </a>
      <a
        class="sc-AxhCb eZTMiI"
        href="/ingredients"
      >
        Ingredients
      </a>
      <a
        class="sc-AxhCb eZTMiI"
        href="/articles"
      >
        Articles
      </a>
      <div
        class="sc-AxheI jYYRbo"
      />
      <div>
        <a
          class="sc-AxhCb eZTMiI"
          href="/login"
        >
          Login
        </a>
        <a
          class="sc-AxhCb eZTMiI"
          href="/register"
        >
          Register
        </a>
      </div>
    </nav>
  </header>
</DocumentFragment>
`;

exports[`Header component renders user navigation 1`] = `
<DocumentFragment>
  <header
    class="sc-AxhUy koLwTO"
  >
    <nav
      class="sc-AxgMl fOKyzp"
    >
      <a
        class="sc-AxhCb eZTMiI"
        href="/"
      >
        Home
      </a>
      <a
        class="sc-AxhCb eZTMiI"
        href="/recipes"
      >
        Recipes
      </a>
      <a
        class="sc-AxhCb eZTMiI"
        href="/ingredients"
      >
        Ingredients
      </a>
      <a
        class="sc-AxhCb eZTMiI"
        href="/articles"
      >
        Articles
      </a>
      <div
        class="sc-AxheI jYYRbo"
      />
      <div />
    </nav>
  </header>
</DocumentFragment>
`;
